cpp/src/arrow/compute/exec/hash_join.cc

```diff
@@ -103,7 +103,8 @@ class HashJoinBasicImpl : public HashJoinImpl {
     filter_ = std::move(filter);
     output_batch_callback_ = std::move(output_batch_callback);
     finished_callback_ = std::move(finished_callback);
-    local_states_.resize(num_threads + 1);  // +1 for calling thread + worker threads
+    // Adding ten, since each side of join might have an IO thread being called from.
+    local_states_.resize(num_threads + 10);
```

Comment:
```suggestion
    local_states_.resize(GetCpuThreadPoolCapacity() + io::GetIOThreadPoolCapacity() + 1);
```
This is slightly less arbitrary than 10 and we'd have to be doing something truly weird to go over this value.